### ISSUE TITLE
add-layer (from wms) should be postponed until capabilities has loaded

### ISSUE THREAD

**Issue by pvgenuchten** [TerriaJS/terriajs#6865]
Some wms servers unfortunately are quite slow in returning response to getcapabilities

If a user tries to add a layer while capabilities is still loading, an error is shown:

![image](https://github.com/TerriaJS/terriajs/assets/299829/4664bd14-fbb0-4d46-a920-df87e62325ea)

Wait a few seconds and try again, it works fine

suggestion would be to disable the panel until capabilities has loaded (now only the spinner is shown)

---

this problem also occurs when layers are loaded by default (via workbench configuration), terria tries to add them while capabilities is still loading, causing the same error

---

An example of a slow capabilities is: https://gaez-services.fao.org/server/services/LR/ImageServer/WMSServer?request=GetCapabilities&service=WMS

**Comment by nf-s** [comment]
Hi @pvgenuchten 

You shouldn't be able to add a WMS layer until GetCapabilities has finished loading.

Can you please share your catalog/init JSON file and/or the process you took to get the error?

Thanks
Nick

**Comment by pvgenuchten** [comment]
There are 2 scenario's in which this error occurs

- a config includes a wms-group with some predefined members, in catalogue open the wms-group, and directly click the first member (while the getcapabilities is still running), see [gaez.zip](https://github.com/TerriaJS/terriajs/files/12511078/gaez.zip)
- if i load a config (or shared map) which has a wms-layer enabled via workbench configuration, see this [shared map](http://localhost/#start=%7B%22version%22%3A%228.0.0%22%2C%22initSources%22%3A%5B%7B%22stratum%22%3A%22user%22%2C%22models%22%3A%7B%22Wapor%22%3A%7B%22isOpen%22%3Atrue%2C%22knownContainerUniqueIds%22%3A%5B%22%2F%22%5D%2C%22type%22%3A%22wms-group%22%7D%2C%22Wapor%2FWAPOR_2%3Al1_aeti_a%22%3A%7B%22knownContainerUniqueIds%22%3A%5B%22Wapor%22%5D%2C%22type%22%3A%22wms%22%7D%2C%22GAEZ%22%3A%7B%22isOpen%22%3Atrue%2C%22knownContainerUniqueIds%22%3A%5B%22%2F%22%5D%2C%22type%22%3A%22wms-group%22%7D%2C%22GAEZ%2FLR%3ADominant+Soil%22%3A%7B%22currentTime%22%3A%222023-09-03T00%3A00%3A00.000000000Z%22%2C%22isPaused%22%3Atrue%2C%22knownContainerUniqueIds%22%3A%5B%22GAEZ%22%5D%2C%22type%22%3A%22wms%22%7D%2C%22Catalogue-kenya%22%3A%7B%22isOpen%22%3Atrue%2C%22knownContainerUniqueIds%22%3A%5B%22%2F%22%5D%2C%22type%22%3A%22csw-group%22%7D%2C%22Catalogue-kenya%2Fdigitalearth-s2_l2a%22%3A%7B%22show%22%3Afalse%2C%22currentTime%22%3A%222023-09-03T00%3A00%3A00.000000000Z%22%2C%22startTime%22%3A%221981-01-01T00%3A00%3A00.000000000Z%22%2C%22stopTime%22%3A%222023-09-03T00%3A00%3A00.000000000Z%22%2C%22multiplier%22%3A8608.29551731192%2C%22isPaused%22%3Atrue%2C%22knownContainerUniqueIds%22%3A%5B%22Catalogue-kenya%22%5D%2C%22type%22%3A%22wms%22%7D%2C%22%2F%22%3A%7B%22type%22%3A%22group%22%7D%7D%2C%22workbench%22%3A%5B%22Wapor%2FWAPOR_2%3Al1_aeti_a%22%5D%2C%22timeline%22%3A%5B%22Wapor%2FWAPOR_2%3Al1_aeti_a%22%5D%2C%22initialCamera%22%3A%7B%22west%22%3A10.944290406450962%2C%22south%22%3A-0.9216353861562537%2C%22east%22%3A36.56440759395097%2C%22north%22%3A14.818507925002162%7D%2C%22homeCamera%22%3A%7B%22west%22%3A-20%2C%22south%22%3A-35%2C%22east%22%3A51%2C%22north%22%3A37%7D%2C%22viewerMode%22%3A%222d%22%2C%22showSplitter%22%3Afalse%2C%22splitPosition%22%3A0.4999%2C%22settings%22%3A%7B%22baseMaximumScreenSpaceError%22%3A2%2C%22useNativeResolution%22%3Afalse%2C%22alwaysShowTimeline%22%3Afalse%2C%22baseMapId%22%3A%22basemap-positron%22%2C%22terrainSplitDirection%22%3A0%2C%22depthTestAgainstTerrainEnabled%22%3Afalse%7D%2C%22stories%22%3A%5B%5D%7D%5D%7D), with terria running at localhost:80

error displayed is 'Could not load the Web Map Service (WMS) GetCapabilities document because the catalog item does not have a url', but my impression is that it is related to gecapabilites not finished loading

**Comment by pvgenuchten** [comment]
The shared map in second bullet may actually be another problem, I just noticed that the share map link does not include a reference to the wms, it refers to a map config having that url, maybe that is somehow related to the error. 

Because if i try this [shared map](http://localhost/#start=%7B%22version%22%3A%228.0.0%22%2C%22initSources%22%3A%5B%7B%22stratum%22%3A%22user%22%2C%22models%22%3A%7B%22__User-Added_Data__%22%3A%7B%22isOpen%22%3Atrue%2C%22members%22%3A%5B%22%2Fhttps%3A%2F%2Fio.apps.fao.org%2Fgeoserver%2Fwms%3Fnamespace%3DWAPOR_2%26service%3Dwms%26request%3DGetCapabilities%22%5D%2C%22knownContainerUniqueIds%22%3A%5B%22%2F%22%5D%2C%22type%22%3A%22group%22%7D%2C%22%2Fhttps%3A%2F%2Fio.apps.fao.org%2Fgeoserver%2Fwms%3Fnamespace%3DWAPOR_2%26service%3Dwms%26request%3DGetCapabilities%22%3A%7B%22isOpen%22%3Atrue%2C%22url%22%3A%22https%3A%2F%2Fio.apps.fao.org%2Fgeoserver%2Fwms%3Fnamespace%3DWAPOR_2%26service%3Dwms%26request%3DGetCapabilities%22%2C%22knownContainerUniqueIds%22%3A%5B%22__User-Added_Data__%22%5D%2C%22type%22%3A%22wms-group%22%7D%2C%22%2Fhttps%3A%2F%2Fio.apps.fao.org%2Fgeoserver%2Fwms%3Fnamespace%3DWAPOR_2%26service%3Dwms%26request%3DGetCapabilities%2FWAPOR_2%3Al1_aeti_d%22%3A%7B%22knownContainerUniqueIds%22%3A%5B%22%2Fhttps%3A%2F%2Fio.apps.fao.org%2Fgeoserver%2Fwms%3Fnamespace%3DWAPOR_2%26service%3Dwms%26request%3DGetCapabilities%22%5D%2C%22type%22%3A%22wms%22%7D%2C%22%2F%22%3A%7B%22type%22%3A%22group%22%7D%7D%2C%22workbench%22%3A%5B%22%2Fhttps%3A%2F%2Fio.apps.fao.org%2Fgeoserver%2Fwms%3Fnamespace%3DWAPOR_2%26service%3Dwms%26request%3DGetCapabilities%2FWAPOR_2%3Al1_aeti_d%22%5D%2C%22timeline%22%3A%5B%22%2Fhttps%3A%2F%2Fio.apps.fao.org%2Fgeoserver%2Fwms%3Fnamespace%3DWAPOR_2%26service%3Dwms%26request%3DGetCapabilities%2FWAPOR_2%3Al1_aeti_d%22%5D%2C%22initialCamera%22%3A%7B%22west%22%3A-78.92578125000001%2C%22south%22%3A-43.45291889355465%2C%22east%22%3A110.03906250000001%2C%22north%22%3A45.33670190996814%7D%2C%22homeCamera%22%3A%7B%22west%22%3A-20%2C%22south%22%3A-35%2C%22east%22%3A51%2C%22north%22%3A37%7D%2C%22viewerMode%22%3A%222d%22%2C%22showSplitter%22%3Afalse%2C%22splitPosition%22%3A0.4999%2C%22settings%22%3A%7B%22baseMaximumScreenSpaceError%22%3A2%2C%22useNativeResolution%22%3Afalse%2C%22alwaysShowTimeline%22%3Afalse%2C%22baseMapId%22%3A%22basemap-positron%22%2C%22terrainSplitDirection%22%3A0%2C%22depthTestAgainstTerrainEnabled%22%3Afalse%7D%2C%22stories%22%3A%5B%5D%7D%5D%7D) it works fine

**Comment by nf-s** [comment]
Oh I see. Yes I agree, members shouldn't show until the group has finished loading. Because you have defined manual `wms` items, Terria won't have "merged" them with items generated by the `GetCapabilities` response - so until it finishes loading they will be incomplete.

I made an issue to capture this 

- https://github.com/TerriaJS/terriajs/issues/6867

We generally discourage use of `members` in "dynamic" groups like `wms-group`, as it can result in "stub"/incomplete items - for example if a WMS server were to unpublish a layer, or if a network error were to occur, you would be left with an incomplete WMS item (in the same way as you found before loading had finished).

What are you trying to achieve by manually defining `wms-group` `members`? There are a few [other properties](https://docs.terria.io/guide/connecting-to-data/catalog-type-details/wms-group/#webmapservicecataloggrouptraits) which may be useful - for example:

`flatten`
> True to flatten the layers into a single list; false to use the layer hierarchy.

`excludeMembers`
> An array of strings of excluded group and item names (or ids).

`itemPropertiesByIds`
> Sets traits on group member items by model ID.

For example - this will override the name of a `GAEZ/LR:Dominant Soil` and exclude `GAEZ/LR:AEZ Classification (33 Classes)`

```json
 {
  "catalog": [
    {
      "id": "GAEZ",
      "type": "wms-group",
      "name": "GAEZ",
      "url": "https://gaez-services.fao.org/server/services/LR/ImageServer/WMSServer?request=GetCapabilities&service=WMS",
      "itemProperties": {
        "rectangle": {
          "west": -180,
          "south": -90,
          "east": 180,
          "north": 90
        }
      },
      "excludeMembers": [
        "GAEZ/LR:AEZ Classification (33 Classes)"
      ],
      "itemPropertiesByIds": [
        {
          "ids": [
            "GAEZ/LR:Dominant Soil"
          ],
          "itemProperties": {
            "name": "Some overridden name"
          }
        }
      ]
    }
  ]
}
```

I believe your second issue is similar to the first - I made an issue to capture this as well

- https://github.com/TerriaJS/terriajs/issues/6868

As for resolving your problem in the short-term - it would probably be easiest to just fully define any `member` that you put inside your `wms-group` (eg include `url`, `layers`, etc).
